File: study/examples/Algorithm/cpp/25_network_flow.cpp
/*
 * 네트워크 플로우 (Network Flow)
 * Ford-Fulkerson, Edmonds-Karp, Dinic, Bipartite Matching
 *
 * 그래프에서 최대 유량을 구하는 알고리즘입니다.
 */

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <climits>
#include <cstring>

using namespace std;

const int INF = INT_MAX;

// =============================================================================
// 1. Ford-Fulkerson (DFS)
// =============================================================================

class FordFulkerson {
private:
    int n;
    vector<vector<int>> capacity;
    vector<vector<int>> adj;

    int dfs(int u, int t, int pushed, vector<bool>& visited) {
        if (u == t) return pushed;

        visited[u] = true;

        for (int v : adj[u]) {
            if (!visited[v] && capacity[u][v] > 0) {
                int flow = dfs(v, t, min(pushed, capacity[u][v]), visited);
                if (flow > 0) {
                    capacity[u][v] -= flow;
                    capacity[v][u] += flow;
                    return flow;
                }
            }
        }

        return 0;
    }

public:
    FordFulkerson(int n) : n(n), capacity(n, vector<int>(n, 0)), adj(n) {}

    void addEdge(int u, int v, int cap) {
        capacity[u][v] += cap;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    int maxFlow(int s, int t) {
        int flow = 0;

        while (true) {
            vector<bool> visited(n, false);
            int pushed = dfs(s, t, INF, visited);
            if (pushed == 0) break;
            flow += pushed;
        }

        return flow;
    }
};

// =============================================================================
// 2. Edmonds-Karp (BFS)
// =============================================================================

class EdmondsKarp {
private:
    int n;
    vector<vector<int>> capacity;
    vector<vector<int>> adj;

public:
    EdmondsKarp(int n) : n(n), capacity(n, vector<int>(n, 0)), adj(n) {}

    void addEdge(int u, int v, int cap) {
        capacity[u][v] += cap;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    int maxFlow(int s, int t) {
        int flow = 0;

        while (true) {
            vector<int> parent(n, -1);
            queue<int> q;
            q.push(s);
            parent[s] = s;

            while (!q.empty() && parent[t] == -1) {
                int u = q.front();
                q.pop();

                for (int v : adj[u]) {
                    if (parent[v] == -1 && capacity[u][v] > 0) {
                        parent[v] = u;
                        q.push(v);
                    }
                }
            }

            if (parent[t] == -1) break;

            // 경로상 최소 용량
            int pushed = INF;
            for (int v = t; v != s; v = parent[v]) {
                pushed = min(pushed, capacity[parent[v]][v]);
            }

            // 용량 업데이트
            for (int v = t; v != s; v = parent[v]) {
                capacity[parent[v]][v] -= pushed;
                capacity[v][parent[v]] += pushed;
            }

            flow += pushed;
        }

        return flow;
    }
};

// =============================================================================
// 3. Dinic's Algorithm
// =============================================================================

class Dinic {
private:
    struct Edge {
        int to, cap, rev;
    };

    int n;
    vector<vector<Edge>> graph;
    vector<int> level, iter;

    bool bfs(int s, int t) {
        fill(level.begin(), level.end(), -1);
        queue<int> q;
        level[s] = 0;
        q.push(s);

        while (!q.empty()) {
            int v = q.front();
            q.pop();

            for (auto& e : graph[v]) {
                if (e.cap > 0 && level[e.to] < 0) {
                    level[e.to] = level[v] + 1;
                    q.push(e.to);
                }
            }
        }

        return level[t] >= 0;
    }

    int dfs(int v, int t, int f) {
        if (v == t) return f;

        for (int& i = iter[v]; i < (int)graph[v].size(); i++) {
            Edge& e = graph[v][i];
            if (e.cap > 0 && level[v] < level[e.to]) {
                int d = dfs(e.to, t, min(f, e.cap));
                if (d > 0) {
                    e.cap -= d;
                    graph[e.to][e.rev].cap += d;
                    return d;
                }
            }
        }

        return 0;
    }

public:
    Dinic(int n) : n(n), graph(n), level(n), iter(n) {}

    void addEdge(int from, int to, int cap) {
        graph[from].push_back({to, cap, (int)graph[to].size()});
        graph[to].push_back({from, 0, (int)graph[from].size() - 1});
    }

    int maxFlow(int s, int t) {
        int flow = 0;

        while (bfs(s, t)) {
            fill(iter.begin(), iter.end(), 0);
            int f;
            while ((f = dfs(s, t, INF)) > 0) {
                flow += f;
            }
        }

        return flow;
    }
};

// =============================================================================
// 4. 이분 매칭 (Hungarian / Hopcroft-Karp)
// =============================================================================

class BipartiteMatching {
private:
    int n, m;
    vector<vector<int>> adj;
    vector<int> matchL, matchR;
    vector<bool> visited;

    bool dfs(int u) {
        for (int v : adj[u]) {
            if (visited[v]) continue;
            visited[v] = true;

            if (matchR[v] == -1 || dfs(matchR[v])) {
                matchL[u] = v;
                matchR[v] = u;
                return true;
            }
        }
        return false;
    }

public:
    BipartiteMatching(int n, int m) : n(n), m(m), adj(n), matchL(n, -1), matchR(m, -1) {}

    void addEdge(int u, int v) {
        adj[u].push_back(v);
    }

    int maxMatching() {
        int result = 0;

        for (int u = 0; u < n; u++) {
            visited.assign(m, false);
            if (dfs(u)) result++;
        }

        return result;
    }

    vector<pair<int, int>> getMatching() {
        vector<pair<int, int>> result;
        for (int u = 0; u < n; u++) {
            if (matchL[u] != -1) {
                result.push_back({u, matchL[u]});
            }
        }
        return result;
    }
};

// =============================================================================
// 5. 최소 비용 최대 유량 (MCMF)
// =============================================================================

class MCMF {
private:
    struct Edge {
        int to, cap, cost, rev;
    };

    int n;
    vector<vector<Edge>> graph;

public:
    MCMF(int n) : n(n), graph(n) {}

    void addEdge(int from, int to, int cap, int cost) {
        graph[from].push_back({to, cap, cost, (int)graph[to].size()});
        graph[to].push_back({from, 0, -cost, (int)graph[from].size() - 1});
    }

    pair<int, int> minCostMaxFlow(int s, int t) {
        int totalFlow = 0, totalCost = 0;

        while (true) {
            vector<int> dist(n, INF);
            vector<int> prevV(n, -1), prevE(n, -1);
            vector<bool> inQueue(n, false);
            queue<int> q;

            dist[s] = 0;
            q.push(s);
            inQueue[s] = true;

            while (!q.empty()) {
                int v = q.front();
                q.pop();
                inQueue[v] = false;

                for (int i = 0; i < (int)graph[v].size(); i++) {
                    Edge& e = graph[v][i];
                    if (e.cap > 0 && dist[v] + e.cost < dist[e.to]) {
                        dist[e.to] = dist[v] + e.cost;
                        prevV[e.to] = v;
                        prevE[e.to] = i;
                        if (!inQueue[e.to]) {
                            q.push(e.to);
                            inQueue[e.to] = true;
                        }
                    }
                }
            }

            if (dist[t] == INF) break;

            // 경로상 최소 용량
            int flow = INF;
            for (int v = t; v != s; v = prevV[v]) {
                flow = min(flow, graph[prevV[v]][prevE[v]].cap);
            }

            // 용량 업데이트
            for (int v = t; v != s; v = prevV[v]) {
                Edge& e = graph[prevV[v]][prevE[v]];
                e.cap -= flow;
                graph[v][e.rev].cap += flow;
            }

            totalFlow += flow;
            totalCost += flow * dist[t];
        }

        return {totalFlow, totalCost};
    }
};

// =============================================================================
// 6. 최소 컷
// =============================================================================

vector<pair<int, int>> minCut(int n, vector<vector<int>>& capacity, int s, int t) {
    // Edmonds-Karp 실행 후 잔여 그래프에서 s에서 도달 가능한 정점 찾기
    vector<vector<int>> residual = capacity;
    vector<vector<int>> adj(n);

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (capacity[i][j] > 0) {
                adj[i].push_back(j);
                adj[j].push_back(i);
            }
        }
    }

    // Max flow
    while (true) {
        vector<int> parent(n, -1);
        queue<int> q;
        q.push(s);
        parent[s] = s;

        while (!q.empty() && parent[t] == -1) {
            int u = q.front();
            q.pop();

            for (int v : adj[u]) {
                if (parent[v] == -1 && residual[u][v] > 0) {
                    parent[v] = u;
                    q.push(v);
                }
            }
        }

        if (parent[t] == -1) break;

        int pushed = INF;
        for (int v = t; v != s; v = parent[v]) {
            pushed = min(pushed, residual[parent[v]][v]);
        }

        for (int v = t; v != s; v = parent[v]) {
            residual[parent[v]][v] -= pushed;
            residual[v][parent[v]] += pushed;
        }
    }

    // BFS로 s에서 도달 가능한 정점
    vector<bool> reachable(n, false);
    queue<int> q;
    q.push(s);
    reachable[s] = true;

    while (!q.empty()) {
        int u = q.front();
        q.pop();

        for (int v : adj[u]) {
            if (!reachable[v] && residual[u][v] > 0) {
                reachable[v] = true;
                q.push(v);
            }
        }
    }

    // 최소 컷 간선
    vector<pair<int, int>> cut;
    for (int u = 0; u < n; u++) {
        if (reachable[u]) {
            for (int v = 0; v < n; v++) {
                if (!reachable[v] && capacity[u][v] > 0) {
                    cut.push_back({u, v});
                }
            }
        }
    }

    return cut;
}

// =============================================================================
// 테스트
// =============================================================================

int main() {
    cout << "============================================================" << endl;
    cout << "네트워크 플로우 예제" << endl;
    cout << "============================================================" << endl;

    // 1. Ford-Fulkerson
    cout << "\n[1] Ford-Fulkerson (DFS)" << endl;
    FordFulkerson ff(6);
    ff.addEdge(0, 1, 16);
    ff.addEdge(0, 2, 13);
    ff.addEdge(1, 2, 10);
    ff.addEdge(1, 3, 12);
    ff.addEdge(2, 1, 4);
    ff.addEdge(2, 4, 14);
    ff.addEdge(3, 2, 9);
    ff.addEdge(3, 5, 20);
    ff.addEdge(4, 3, 7);
    ff.addEdge(4, 5, 4);
    cout << "    최대 유량 (0 → 5): " << ff.maxFlow(0, 5) << endl;

    // 2. Edmonds-Karp
    cout << "\n[2] Edmonds-Karp (BFS)" << endl;
    EdmondsKarp ek(6);
    ek.addEdge(0, 1, 16);
    ek.addEdge(0, 2, 13);
    ek.addEdge(1, 2, 10);
    ek.addEdge(1, 3, 12);
    ek.addEdge(2, 1, 4);
    ek.addEdge(2, 4, 14);
    ek.addEdge(3, 2, 9);
    ek.addEdge(3, 5, 20);
    ek.addEdge(4, 3, 7);
    ek.addEdge(4, 5, 4);
    cout << "    최대 유량 (0 → 5): " << ek.maxFlow(0, 5) << endl;

    // 3. Dinic
    cout << "\n[3] Dinic's Algorithm" << endl;
    Dinic dinic(6);
    dinic.addEdge(0, 1, 16);
    dinic.addEdge(0, 2, 13);
    dinic.addEdge(1, 2, 10);
    dinic.addEdge(1, 3, 12);
    dinic.addEdge(2, 1, 4);
    dinic.addEdge(2, 4, 14);
    dinic.addEdge(3, 2, 9);
    dinic.addEdge(3, 5, 20);
    dinic.addEdge(4, 3, 7);
    dinic.addEdge(4, 5, 4);
    cout << "    최대 유량 (0 → 5): " << dinic.maxFlow(0, 5) << endl;

    // 4. 이분 매칭
    cout << "\n[4] 이분 매칭" << endl;
    BipartiteMatching bm(4, 4);
    bm.addEdge(0, 0);
    bm.addEdge(0, 1);
    bm.addEdge(1, 0);
    bm.addEdge(1, 2);
    bm.addEdge(2, 1);
    bm.addEdge(2, 2);
    bm.addEdge(3, 2);
    bm.addEdge(3, 3);
    cout << "    최대 매칭: " << bm.maxMatching() << endl;
    cout << "    매칭: ";
    for (auto [l, r] : bm.getMatching()) {
        cout << "(" << l << "," << r << ") ";
    }
    cout << endl;

    // 5. MCMF
    cout << "\n[5] 최소 비용 최대 유량" << endl;
    MCMF mcmf(4);
    mcmf.addEdge(0, 1, 2, 1);
    mcmf.addEdge(0, 2, 1, 2);
    mcmf.addEdge(1, 2, 1, 1);
    mcmf.addEdge(1, 3, 1, 3);
    mcmf.addEdge(2, 3, 2, 1);
    auto [flow, cost] = mcmf.minCostMaxFlow(0, 3);
    cout << "    최대 유량: " << flow << ", 최소 비용: " << cost << endl;

    // 6. 복잡도 요약
    cout << "\n[6] 복잡도 요약" << endl;
    cout << "    | 알고리즘       | 시간복잡도      | 특징              |" << endl;
    cout << "    |----------------|-----------------|-------------------|" << endl;
    cout << "    | Ford-Fulkerson | O(Ef)           | DFS, 무한루프 가능|" << endl;
    cout << "    | Edmonds-Karp   | O(VE²)          | BFS, 안정적       |" << endl;
    cout << "    | Dinic          | O(V²E)          | 레벨 그래프       |" << endl;
    cout << "    | 이분 매칭      | O(VE)           | 헝가리안          |" << endl;
    cout << "    | MCMF           | O(VEf) or O(V³E)| SPFA/Bellman-Ford |" << endl;

    cout << "\n============================================================" << endl;

    return 0;
}

File: study/examples/Algorithm/python/25_network_flow.py
"""
네트워크 플로우 (Network Flow)
Maximum Flow and Bipartite Matching

그래프에서 최대 유량을 구하는 알고리즘입니다.
"""

from typing import List, Tuple, Dict
from collections import defaultdict, deque


# =============================================================================
# 1. Ford-Fulkerson (DFS 기반)
# =============================================================================

def ford_fulkerson(capacity: List[List[int]], source: int, sink: int) -> int:
    """
    Ford-Fulkerson 알고리즘 (DFS)
    시간복잡도: O(E * max_flow)
    """
    n = len(capacity)
    residual = [row[:] for row in capacity]

    def dfs(node: int, flow: int, visited: List[bool]) -> int:
        if node == sink:
            return flow

        visited[node] = True

        for next_node in range(n):
            if not visited[next_node] and residual[node][next_node] > 0:
                min_flow = min(flow, residual[node][next_node])
                result = dfs(next_node, min_flow, visited)

                if result > 0:
                    residual[node][next_node] -= result
                    residual[next_node][node] += result
                    return result

        return 0

    max_flow = 0
    while True:
        visited = [False] * n
        flow = dfs(source, float('inf'), visited)
        if flow == 0:
            break
        max_flow += flow

    return max_flow


# =============================================================================
# 2. Edmonds-Karp (BFS 기반)
# =============================================================================

def edmonds_karp(capacity: List[List[int]], source: int, sink: int) -> int:
    """
    Edmonds-Karp 알고리즘 (BFS)
    시간복잡도: O(V * E²)
    """
    n = len(capacity)
    residual = [row[:] for row in capacity]

    def bfs() -> List[int]:
        """BFS로 증가 경로 찾기"""
        parent = [-1] * n
        visited = [False] * n
        visited[source] = True
        queue = deque([source])

        while queue:
            node = queue.popleft()

            for next_node in range(n):
                if not visited[next_node] and residual[node][next_node] > 0:
                    visited[next_node] = True
                    parent[next_node] = node
                    queue.append(next_node)

                    if next_node == sink:
                        return parent

        return parent

    max_flow = 0

    while True:
        parent = bfs()

        if parent[sink] == -1:
            break

        # 경로의 최소 용량 찾기
        path_flow = float('inf')
        node = sink
        while node != source:
            prev = parent[node]
            path_flow = min(path_flow, residual[prev][node])
            node = prev

        # 잔여 그래프 업데이트
        node = sink
        while node != source:
            prev = parent[node]
            residual[prev][node] -= path_flow
            residual[node][prev] += path_flow
            node = prev

        max_flow += path_flow

    return max_flow


# =============================================================================
# 3. Dinic 알고리즘
# =============================================================================

class Dinic:
    """
    Dinic 알고리즘
    시간복잡도: O(V² * E)
    이분 그래프에서: O(E * √V)
    """

    def __init__(self, n: int):
        self.n = n
        self.graph = defaultdict(list)

    def add_edge(self, u: int, v: int, cap: int):
        """간선 추가 (u → v, 용량 cap)"""
        # (인접 노드, 잔여 용량, 역방향 간선 인덱스)
        self.graph[u].append([v, cap, len(self.graph[v])])
        self.graph[v].append([u, 0, len(self.graph[u]) - 1])

    def bfs(self, source: int, sink: int) -> bool:
        """레벨 그래프 구성"""
        self.level = [-1] * self.n
        self.level[source] = 0
        queue = deque([source])

        while queue:
            node = queue.popleft()
            for next_node, cap, _ in self.graph[node]:
                if cap > 0 and self.level[next_node] < 0:
                    self.level[next_node] = self.level[node] + 1
                    queue.append(next_node)

        return self.level[sink] >= 0

    def dfs(self, node: int, sink: int, flow: int) -> int:
        """blocking flow 찾기"""
        if node == sink:
            return flow

        for i in range(self.iter[node], len(self.graph[node])):
            self.iter[node] = i
            next_node, cap, rev = self.graph[node][i]

            if cap > 0 and self.level[next_node] == self.level[node] + 1:
                d = self.dfs(next_node, sink, min(flow, cap))

                if d > 0:
                    self.graph[node][i][1] -= d
                    self.graph[next_node][rev][1] += d
                    return d

        return 0

    def max_flow(self, source: int, sink: int) -> int:
        """최대 유량 계산"""
        flow = 0

        while self.bfs(source, sink):
            self.iter = [0] * self.n

            while True:
                f = self.dfs(source, sink, float('inf'))
                if f == 0:
                    break
                flow += f

        return flow


# =============================================================================
# 4. 이분 매칭 (Bipartite Matching)
# =============================================================================

def bipartite_matching(n: int, m: int, edges: List[Tuple[int, int]]) -> int:
    """
    이분 그래프 최대 매칭 (Hopcroft-Karp 간소화)
    n: 왼쪽 정점 수, m: 오른쪽 정점 수
    edges: [(왼쪽, 오른쪽), ...]
    시간복잡도: O(E * √V)
    """
    # 그래프 구성
    adj = defaultdict(list)
    for u, v in edges:
        adj[u].append(v)

    match_left = [-1] * n   # 왼쪽 정점의 매칭
    match_right = [-1] * m  # 오른쪽 정점의 매칭

    def dfs(u: int, visited: List[bool]) -> bool:
        for v in adj[u]:
            if visited[v]:
                continue
            visited[v] = True

            # 매칭 안됨 또는 재매칭 가능
            if match_right[v] == -1 or dfs(match_right[v], visited):
                match_left[u] = v
                match_right[v] = u
                return True

        return False

    matching = 0
    for u in range(n):
        visited = [False] * m
        if dfs(u, visited):
            matching += 1

    return matching


# =============================================================================
# 5. 최소 컷 (Minimum Cut)
# =============================================================================

def min_cut(capacity: List[List[int]], source: int, sink: int) -> Tuple[int, List[int]]:
    """
    최소 컷 = 최대 유량
    반환: (컷 용량, source 측 정점 집합)
    """
    n = len(capacity)
    residual = [row[:] for row in capacity]

    # Edmonds-Karp로 최대 유량 계산
    def bfs_flow() -> bool:
        parent = [-1] * n
        visited = [False] * n
        visited[source] = True
        queue = deque([source])

        while queue:
            node = queue.popleft()
            for next_node in range(n):
                if not visited[next_node] and residual[node][next_node] > 0:
                    visited[next_node] = True
                    parent[next_node] = node
                    queue.append(next_node)

        if parent[sink] == -1:
            return False

        path_flow = float('inf')
        node = sink
        while node != source:
            prev = parent[node]
            path_flow = min(path_flow, residual[prev][node])
            node = prev

        node = sink
        while node != source:
            prev = parent[node]
            residual[prev][node] -= path_flow
            residual[node][prev] += path_flow
            node = prev

        return True

    while bfs_flow():
        pass

    # 잔여 그래프에서 source에서 도달 가능한 정점 찾기
    visited = [False] * n
    queue = deque([source])
    visited[source] = True

    while queue:
        node = queue.popleft()
        for next_node in range(n):
            if not visited[next_node] and residual[node][next_node] > 0:
                visited[next_node] = True
                queue.append(next_node)

    # 컷 용량 계산
    cut_capacity = 0
    source_side = []
    for i in range(n):
        if visited[i]:
            source_side.append(i)
            for j in range(n):
                if not visited[j] and capacity[i][j] > 0:
                    cut_capacity += capacity[i][j]

    return cut_capacity, source_side


# =============================================================================
# 6. 실전 문제: 작업 할당
# =============================================================================

def assign_tasks(workers: int, tasks: int, can_do: List[List[int]]) -> List[int]:
    """
    작업자에게 작업 할당 (이분 매칭)
    can_do[i] = worker i가 수행 가능한 작업 목록
    반환: assignment[worker] = 할당된 작업 (-1이면 미할당)
    """
    edges = []
    for worker, tasks_list in enumerate(can_do):
        for task in tasks_list:
            edges.append((worker, task))

    # 이분 매칭 수행
    adj = defaultdict(list)
    for u, v in edges:
        adj[u].append(v)

    match_worker = [-1] * workers
    match_task = [-1] * tasks

    def dfs(u: int, visited: List[bool]) -> bool:
        for v in adj[u]:
            if visited[v]:
                continue
            visited[v] = True

            if match_task[v] == -1 or dfs(match_task[v], visited):
                match_worker[u] = v
                match_task[v] = u
                return True

        return False

    for u in range(workers):
        visited = [False] * tasks
        dfs(u, visited)

    return match_worker


# =============================================================================
# 7. 실전 문제: 최대 간선 분리 경로
# =============================================================================

def max_edge_disjoint_paths(n: int, edges: List[Tuple[int, int]], source: int, sink: int) -> int:
    """
    간선 분리 경로의 최대 개수
    각 간선 용량 = 1로 설정한 최대 유량
    """
    capacity = [[0] * n for _ in range(n)]

    for u, v in edges:
        capacity[u][v] = 1
        capacity[v][u] = 1  # 무방향 그래프인 경우

    return edmonds_karp(capacity, source, sink)


# =============================================================================
# 테스트
# =============================================================================

def main():
    print("=" * 60)
    print("네트워크 플로우 (Network Flow) 예제")
    print("=" * 60)

    # 그래프 예시:
    #     10      10
    #  0 ───→ 1 ───→ 3
    #  │      │      ↑
    #  │10    │2     │10
    #  ↓      ↓      │
    #  2 ───→ 4 ───→ 5
    #     10      10

    # 1. Ford-Fulkerson
    print("\n[1] Ford-Fulkerson")
    capacity = [
        [0, 10, 10, 0, 0, 0],  # 0
        [0, 0, 2, 10, 0, 0],   # 1
        [0, 0, 0, 0, 10, 0],   # 2
        [0, 0, 0, 0, 0, 10],   # 3
        [0, 0, 0, 0, 0, 10],   # 4
        [0, 0, 0, 0, 0, 0]     # 5 (sink)
    ]
    flow = ford_fulkerson(capacity, 0, 5)
    print(f"    source=0, sink=5")
    print(f"    최대 유량: {flow}")

    # 2. Edmonds-Karp
    print("\n[2] Edmonds-Karp")
    flow = edmonds_karp(capacity, 0, 5)
    print(f"    최대 유량: {flow}")

    # 3. Dinic
    print("\n[3] Dinic 알고리즘")
    dinic = Dinic(6)
    dinic.add_edge(0, 1, 10)
    dinic.add_edge(0, 2, 10)
    dinic.add_edge(1, 2, 2)
    dinic.add_edge(1, 3, 10)
    dinic.add_edge(2, 4, 10)
    dinic.add_edge(3, 5, 10)
    dinic.add_edge(4, 5, 10)
    flow = dinic.max_flow(0, 5)
    print(f"    최대 유량: {flow}")

    # 4. 이분 매칭
    print("\n[4] 이분 매칭")
    # 왼쪽: 0, 1, 2 (작업자)
    # 오른쪽: 0, 1, 2 (작업)
    edges = [(0, 0), (0, 1), (1, 0), (1, 2), (2, 1)]
    matching = bipartite_matching(3, 3, edges)
    print(f"    간선: {edges}")
    print(f"    최대 매칭: {matching}")

    # 5. 최소 컷
    print("\n[5] 최소 컷")
    cut_cap, source_side = min_cut(capacity, 0, 5)
    print(f"    최소 컷 용량: {cut_cap}")
    print(f"    source 측 정점: {source_side}")

    # 6. 작업 할당
    print("\n[6] 작업 할당")
    can_do = [
        [0, 1],     # 작업자 0: 작업 0, 1 가능
        [1, 2],     # 작업자 1: 작업 1, 2 가능
        [0, 2]      # 작업자 2: 작업 0, 2 가능
    ]
    assignment = assign_tasks(3, 3, can_do)
    print(f"    수행 가능: {can_do}")
    print(f"    할당 결과: {assignment}")

    # 7. 알고리즘 비교
    print("\n[7] 알고리즘 복잡도 비교")
    print("    | 알고리즘      | 시간 복잡도       | 특징               |")
    print("    |---------------|-------------------|-------------------|")
    print("    | Ford-Fulkerson| O(E * max_flow)   | DFS, 정수 용량     |")
    print("    | Edmonds-Karp  | O(V * E²)         | BFS, 안정적        |")
    print("    | Dinic         | O(V² * E)         | 레벨 그래프        |")
    print("    | 이분 매칭     | O(E * √V)         | Dinic 특수 케이스  |")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()
